### DISCUSSION TITLE
Facing Issue in the iModel Presentation API (2.19.2)

### DISCUSSION THREAD

**Discussion by pahm3y** [iTwin/itwinjs-core#4447] [General]
We are observing below error during the external attribute extraction using the Presentation APIs:
  **D:\Siemens\Teamcenter14\TC_ROOT\CALM_Utilities\Plantcad2Tc\node_modules\@bentley\presentation-backend\lib\presentation-backend\NativePlatform.js:120
                      const pollResult = this._nativeAddon.pollResponse(requestGuid);
                                                             ^
  Error: Unknown failure
      at Timeout._onTimeout (D:\Siemens\Teamcenter14\TC_ROOT\CALM_Utilities\Plantcad2Tc\node_modules\@bentley\presentation-backend\lib\presentation-backend\NativePlatform.js:120:58)
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)**
we are calling the following API:
 content = await Presentation.getManager().getContent({requestContext: ClientRequestContext.current, imodel: iModel,
      rulesetOrId: ruleSet , descriptor: overrides, keys: keySet});
where ruleset is defined as below:
`const ruleSet: Ruleset = {
      id: `properties`,
      rules: [{
        ruleType: RuleTypes.Content,
        specifications: [{ 
          specType: ContentSpecificationTypes.SelectedNodeInstances,
          acceptableSchemaName: schemaName,
          acceptableClassNames: classNames,
          acceptablePolymorphically: true,
        }]
      }],
    };`
               _"schemaName = PDMxPlant_Chevron"
"classNames = [  "ATTACHMENT",  "BPANEL",  "ELBOW",  "FITTING",  "FLUID_PORT",  "FTUBE",  "GASKET",  "GENSEC",  "NOZZLE",  "PANEL",  "PORT",  "REDUCER",  "SCTN",  "STRUCTURE",  "SUBSTRUCTURE",  "TEE",  "TMPLATE",  "TUBE", "WELD"];
"The count of value in the keyset is 28650."_

`IModelJS library version : 2.19.2`

**Comment by aruniverse** [comment]
Have you tried the latest 2.19.x version?

**Comment by grigasp** [comment]
The error doesn't originate from us, so it's hard to tell what it means without additional details. Could you get a trace level log for the “ECPresentation” logger namespace or provide access to your iModel?

One red flag is that the request is not paged. A content response for nearly 30k elements is going to be a massive JSON, so the first thing I'd try would be to request content in pages. At our RPC layer we force a max page size of 1k.

**Reply by pahm3y** [reply]
What would be the API or settings to request the content in pages. I am using the iModelJS version 2.19.2 and cannot use any other version as of now.

**Reply by grigasp** [reply]
The `requestOptions` you pass to [getContent](https://github.com/iTwin/itwinjs-core/blob/953ec6f0e3f7cf3713aeec4c348866a88d281a4d/presentation/backend/src/presentation-backend/PresentationManager.ts#L707) is [Paged](https://github.com/iTwin/itwinjs-core/blob/953ec6f0e3f7cf3713aeec4c348866a88d281a4d/presentation/common/src/presentation-common/PresentationManagerOptions.ts#L255), which means you can pass `pageOptions` as one of the parameters.
